src/main/java/build/buf/protovalidate/Ipv6.java

```diff
@@ -264,34 +272,39 @@ private boolean dotted() {
    *
    * <pre>h16 = 1*4HEXDIG
    *
-   * <p>Stores 16-bit value in pieces.
+   * If 1-4 hex digits are found, the parsed 16-bit unsigned integer is stored
+   * in pieces and true is returned.
+   * If 0 hex digits are found, returns false.
+   * If more than 4 hex digits are found, a ParseException is thrown.
+   * If the found hex digits cannot be converted to an int, a NumberFormatException is raised.
    */
-  private boolean h16() {
+  private boolean h16() throws ParseException, NumberFormatException {
     int start = this.index;
 
     while (this.index < this.str.length() && this.hexDig()) {}
 
     String str = this.str.substring(start, this.index);
 
     if (str.isEmpty()) {
-      // too short
+      // too short, just return false
+      // this is not an error condition, it just means we didn't find any
+      // hex digits at the current position.
       return false;
     }
 
     if (str.length() > 4) {
       // too long
-      return false;
+      // this is an error condition, it means we found a string of more than
+      // four valid hex digits, which is invalid in ipv6 addresses.
+      throw new ParseException("invalid hex", this.index);
```

Comment:
If we want to use exceptions for control flow here, I'd throw an `IllegalStateException` here over a `ParseException`. 
```suggestion
      throw new IllegalStateException("invalid hex");
```